src/core/mempool.rs

```diff
@@ -956,11 +1114,124 @@ impl MemPoolDB {
         }
     }
 
+    /// Remove all txids at the given height from the bloom counter.
+    /// Used to clear out txids that are now outside the bloom counter's depth.
+    fn prune_bloom_counter(tx: &mut MemPoolTx, target_height: u64) -> Result<(), MemPoolRejection> {
+        let sql = "SELECT txid FROM mempool WHERE height = ?1 AND NOT EXISTS (SELECT 1 FROM removed_txids WHERE txid = mempool.txid)";
+        let args: &[&dyn ToSql] = &[&u64_to_sql(target_height)?];
+        let txids: Vec<Txid> = query_rows(tx, sql, args)?;
+        let num_txs = txids.len();
+
+        debug!("Prune bloom counter from height {}", target_height);
+
+        // keep borrow-checker happy
+        let bloom_counter = tx.take_bloom_state();
+        for txid in txids.into_iter() {
+            bloom_counter.remove_raw(&mut tx.tx, &txid.0)?;
+
+            let sql = "INSERT OR REPLACE INTO removed_txids (txid) VALUES (?1)";
+            let args: &[&dyn ToSql] = &[&txid];
+            tx.execute(sql, args).map_err(db_error::SqliteError)?;
+        }
+
+        debug!(
+            "Pruned bloom filter at height {}: removed {} txs",
+            target_height, num_txs
+        );
+        tx.replace_bloom_state(bloom_counter);
+        Ok(())
+    }
+
+    /// Add the txid to the bloom counter in the mempool DB.
+    /// If this is the first txid at this block height, then also garbage-collect the bloom counter to remove no-longer-recent transactions.
+    /// If the bloom counter is saturated -- i.e. it represents more than MAX_BLOOM_COUNTER_TXS
+    /// transactions -- then pick another transaction to evict from the bloom filter and return its txid.
+    /// (Note that no transactions are ever removed from the mempool; we just don't prioritize them
+    /// in the bloom filter).
+    fn update_bloom_counter(
+        tx: &mut MemPoolTx,
+        height: u64,
+        txid: &Txid,
+        prior_txid: Option<Txid>,
+    ) -> Result<Option<Txid>, MemPoolRejection> {
+        // is this the first-ever txid at this height?
+        let sql = "SELECT 1 FROM mempool WHERE height = ?1";
+        let args: &[&dyn ToSql] = &[&u64_to_sql(height)?];
+        let present: Option<i64> = query_row(tx, sql, args)?;
+        if present.is_none() && height > (BLOOM_COUNTER_DEPTH as u64) {
+            // this is the first-ever tx at this height.
+            // which means, the bloom filter window has advanced.
+            // which means, we need to remove all the txs that are now out of the window.
+            MemPoolDB::prune_bloom_counter(tx, height - (BLOOM_COUNTER_DEPTH as u64))?;
+        }
+
+        // keep borrow-checker happy
+        let bloom_counter = tx.take_bloom_state();
+
```

Comment:
Same comment as above -- most of these uses of `take_bloom_state` are pretty unsafe. Any thrown errors will cause the bloom state to be lost.

Reply:
Fixed